File: ruby/validate_closure.rb
def validate_closure(seq)
  # Check that a string sequence is valid when using {}, [], ().
  # 
  # For example:
  #   > validate_closure('al{fi[nk]el}(st{e}in)') will return true
  #   > validate_closure('()') will return true
  #   > validate_closure('al{fi[nkel}](st{e}in)') will return false
  #   > validate_closure('(') will return false
  #   > validate_closure(']') will return false

  match = {'(' => ')', '{' => '}', '[' => ']'}
  stack = []
  seq.chars.each do |character|
    if match.has_key?(character)
      stack.push(character)
    end
    if match.has_value?(character)
      if stack.empty?
        return false
      end
      last = stack.pop
      if match[last] != character
        return false
      end
    end
  end
  if !stack.empty?
    return false
  end
  return true
end

tests = ['al{fi[nk]el}(st{e}in)', '()', 'al{fi[nkel}](st{e}in)', '(', ']']

tests.each do |test|
  puts "#{test} -> #{validate_closure(test).to_s}"
end

File: python/validate_closure.py
def validate_closure(seq):
    """
    Check that a string sequence is valid when using {}, [], ().
    
    For example:
        > validate_closure('al{fi[nk]el}(st{e}in)') will return True
        > validate_closure('()') will return True
        > validate_closure('al{fi[nkel}](st{e}in)') will return False
        > validate_closure('(') will return False
        > validate_closure(']') will return False
    """
    match = {'(': ')', '{': '}', '[': ']'}
    stack = []
    for character in seq[:]:
        if character in match.keys():
            stack.append(character)
        if character in match.values():
            if not stack:
                return False
            last = stack.pop()
            if match[last] != character:
                return False
    if stack:
        return False
    return True
